V3 sor with swaps

lgahdl opened the pull request:
Adapting getSorSwapPaths to Balancer V3 Pools(Swaps Only)

Commit:
Adding first version of normalized liquidity calculation:
	- Job 'update-on-chain-data-v3'
	- Function to fetch token pair data;
	- Upserting tokenPair into the db;
	- Adding router address and abi;

Commit:
Adapting "getSorSwapPaths" query to work with v3 Pools;
Adding toJSON function in BigInt to fix a problem with BigInt serialization;

Commit:
removing logs;

Commit:
Merge commit '76a2a11d1c701fe0239d16a9d30d96294fb434bd' into v3-sor-with-swaps

# Conflicts:
#	config/sepolia.ts
#	modules/controllers/jobs-controller.ts
#	modules/sor/sorV2/sorPathService.ts
#	tasks/index.ts

Commit:
Merge commit '76a2a11d1c701fe0239d16a9d30d96294fb434bd' into v3-sor-with-swaps

# Conflicts:
#	config/sepolia.ts
#	modules/controllers/jobs-controller.ts
#	modules/sor/sorV2/sorPathService.ts
#	tasks/index.ts

Commit:
Merge commit '1298eaa94ad61d205721b71f54f52be87d7547fd' into v3-sor-with-swaps

# Conflicts:
#	modules/controllers/jobs-controller.ts
#	modules/network/network-config-types.ts
#	modules/pool/lib/pool-on-chain-tokenpair-data.ts
#	modules/sor/sor.service.ts
#	tasks/index.ts

Commit:
Merge commit '76a2a11d1c701fe0239d16a9d30d96294fb434bd' into v3-sor-with-swaps

# Conflicts:
#	config/sepolia.ts
#	modules/controllers/jobs-controller.ts
#	modules/sor/sorV2/sorPathService.ts
#	tasks/index.ts

Commit:
Fixing tokenpair data update on DB;

Commit:
Merge commit 'b01bd5096419b5dd9326a4fdf965cce852b26441' into v3-sor-with-swaps

Commit:
Fixing wrong values on swapPaths return:
 - tokenOutAmount variable;
 - tokenOut variable;

Commit:
Adding normalized liquidity calculation for token pairs with BPT included(Add/Remove Liquidity transactions);

Commit:
running prettier on file;

Commit:
Removing unused file;

Commit:
Merge branch 'v3-sor-with-swaps' into normalized-liquidity-bpts

Commit:
Adding pools as graph nodes;

Commit:
Updating @balancer/sdk package to v0.13.0;

Commit:
Merge branch 'v3-sor-with-swaps' into normalized-liquidity-bpts

Commit:
Merge branch 'normalized-liquidity-bpts' into v3-sor-add-remove-liq

lgahdl committed:
Update modules/sources/contracts/fetch-tokenpair-data.ts

Co-authored-by: Bruno <bruno@balancerlabs.dev>

Commit:
Building Paths considering Add/Remove Liquidity Operations;

Commit:
Updating Balancer SDK package to 0.14.0;
Adding filter to V3 pool types compatible with add/remove liquidity paths;
Adapting swap query output to V0.14.0 query output;

Commit:
Merge commit '4905b2c4e06a51a5bfc6eea2fd57b45c7234c6df' into v3-sor-add-remove-liq

Commit:
Merge commit 'acac63e1883a655cfca29d04b905c701883a9142' into v3-sor-add-remove-liq

Commit:
Removing BigInt polyfill;

Commit:
Merge branch 'v3-sor-with-swaps' into normalized-liquidity-bpts

Commit:
Merge branch 'normalized-liquidity-bpts' into v3-sor-add-remove-liq

Commit:
Removing statement that will always be false (V3 Pools will not have BPT);

Commit:
Merge branch 'normalized-liquidity-bpts' into v3-sor-add-remove-liq

Commit:
Merge commit 'acac63e1883a655cfca29d04b905c701883a9142' into v3-sor-with-swaps

Commit:
Merge branch 'v3-sor-with-swaps' into normalized-liquidity-bpts

Commit:
Merge branch 'normalized-liquidity-bpts' into v3-sor-add-remove-liq

Commit:
Removing log;

Commit:
Merge pull request #297 from balancer/v3-sor-add-remove-liq

V3 SOR - Adapting V3 SOR to build paths with Add/Remove Liquidity Operations;

Commit:
Merge pull request #269 from balancer/normalized-liquidity-bpts

Adding normalized liquidity calculation for token pairs with BPT incl…

franzns commented on modules/sources/contracts/fetch-tokenpair-data.ts:
```diff
@@ -193,6 +218,52 @@ function generateTokenPairs(filteredPools: PoolInput[]): TokenPair[] {
     return tokenPairs;
 }
 
+function generateBptTokenPairs(filteredPools: PoolInput[]): TokenPair[] {
```
I would actually do that in the generateTokenPairs function, you should have all the info needed.

franzns commented on modules/sources/contracts/fetch-tokenpair-data.ts:
```diff
@@ -99,11 +103,27 @@ export async function fetchTokenPairData(
     });
 
+    let bptTokenPairs = generateBptTokenPairs(pools);
+    bptTokenPairs.forEach((bptTokenPair) => {
```
No need to do it separately, you could check in the `addEffectivePriceCallsToMulticaller`, etc. functions whether token A is the pool address (maybe need to add pool address to the TokenPair type)

franzns commented on modules/sources/contracts/fetch-tokenpair-data.ts:
```diff
@@ -255,6 +356,25 @@ function addBToAPriceCallsToMulticaller(
+    balancerRouterAddress: string,
+    multicaller: ViemMulticallCall[],
+) {
+    if (tokenPair.tokenBIndex === undefined || tokenPair.poolTokensLength === undefined) {
```
I would add vaultVersion to the tokenPair type instead of checking for undefined here

lgahdl replied:
This check is just to ensure to TS that tokenBIndex and poolTokensLength will not be undefined, all the tokenPairs will arrive with tokenBIndex and poolsTokensLength defined at this point, but for non-BPT token pairs these parameters can be undefined, so I made this verification instead of forcing "tokenBIndex as number"

franzns commented on modules/sor/sor.service.ts:
```diff
@@ -17,6 +17,9 @@ import { AllNetworkConfigsKeyedOnChain } from '../network/network-config';
 
 export class SorService {
     async getSorSwapPaths(args: QuerySorGetSwapPathsArgs): Promise<GqlSorGetSwapPaths> {
+        if(!!args.useVaultVersion && args.useVaultVersion!==2 && args.useVaultVersion!==3){
```
Whats wrong with your prettier? This should be formatted

franzns replied:
Actually, this check should be different. If there is no vaultVersion specified, it should find routes on both versions and only return the better one.

brunoguerios replied:
Our initial thoughts on the SDK side was that the user would have to choose between v2 or v3.
If they wanted the best path disregard of vault version, they would have to handle that logic externally.
With that being said, now that the API is serving it, I'd agree that it makes more sense to handle that for the user.

franzns commented on modules/sor/sor.service.ts:
```diff
@@ -59,6 +62,7 @@ export class SorService {
             swapType: args.swapType,
             tokenIn: tokenIn,
             tokenOut: tokenOut,
+            vaultVersion: args.useVaultVersion ?? 2,
```
Should just pass the arg

franzns commented on modules/sor/sorV2/sorPathService.ts:
```diff
@@ -102,6 +106,7 @@ class SorPathService implements SwapService {
                 paths!,
                 input.swapType,
                 input.chain,
+                input.vaultVersion as 2 | 3,
```
could in theory be undefined if we need to query both versions

franzns commented on modules/sor/sorV2/sorPathService.ts:
```diff
@@ -399,18 +425,19 @@ class SorPathService implements SwapService {
                     swapEnabled: true,
                     totalLiquidity: {
-                        gt: 1000,
+                        gte: vaultVersion === 2 ? 1000 : 0,
```
should also do the same TVL limit for v3 pools

lgahdl replied:
Sorry, when I did that, all v3 pools had totalLiquidity<1000, I'll remove this now

franzns commented on modules/sor/sorV2/sorPathService.ts:
```diff
@@ -449,7 +476,7 @@ class SorPathService implements SwapService {
         const tokenOut = path.tokens[path.tokens.length - 1].address;
         const tokenInAmount = formatUnits(path.inputAmount.amount, path.tokens[0].decimals);
-        const tokenOutAmount = formatUnits(path.inputAmount.amount, path.tokens[path.tokens.length - 1].decimals);
+        const tokenOutAmount = formatUnits(path.outputAmount.amount, path.tokens[path.tokens.length - 1].decimals);
```
good catch! thanks!

lgahdl replied:
Thanks!

franzns commented on modules/sor/sorV2/lib/pathGraph/pathGraph.ts:
```diff
@@ -180,6 +188,16 @@ export class PathGraph {
     }
 
+    private addPoolsAsGraphNodes(pools: BasePool[]) {
+        const validPoolTypes = ['WEIGHTED', 'COMPOSABLE_STABLE'];
```
Do we need this limitation?

lgahdl replied:
I added this to avoid Gyro and Fx pools being added as nodes, they have no add/remove liquidity support, so it could cause an error

franzns replied:
But this is for v3 only, right? Will gyro and fx pools in v3 also not allow that?

brunoguerios replied:
Franz is right. Swaps with add/remove liquidity is for v3 only and pools will have support for all add/remove liquidity types out of the box.

lgahdl replied:
Ok, I will need to add the pool math for these operations for these 2 pool types to calculate limits

brunoguerios replied:
Maybe add a `TODO` comment explaining that this is temporary until we start adding other pool types for v3?

franzns commented on modules/sor/sorV2/lib/pools/weighted/weightedPool.ts:
```diff
@@ -64,17 +67,13 @@ export class WeightedPool implements BasePool {
-                    poolToken.index,
-                ),
-            );
+            //TODO: Remove this once the weight for V3 Pools is in the same format as V2 Pools
```
@gmbronco seems we messed up tokenweight types for v3, need to fix

franzns commented:
Thanks for the PR. A few good catches and fixes in there :)

General remark:
1) The idea was, that if one does not supply a vault version, it would query both v2 and v3 and return the better paths. Can you adapt to code for that or do we plan to add it later?
2) I would try to do the tokenPair data fetching in one go, whether they are BPT pairs or token Pairs. 
~~3) Can you add poolToken->Phantom BPT pairs for CSP for v2? As @brunoguerios found out, there is a bug. CSP v2 pools support swap from pool token to phantom BPT.~~ Bruno will add it to his PR
4) @brunoguerios a lot of pool math made its way into the API now. I wonder if we should start the project of the "pool math repo". 

I havent reviewed all the math stuff, did you review that @brunoguerios ?

Did you run any queries and checked if it works? @brunoguerios @lgahdl 😅

brunoguerios commented:
Answering your question about the `pool math repo`, we've talked a bit about how it fits in the SDK/Integrations roadmap.
It will likely be something useful for aggregators, which should be one of our main priorities as we move towards an Integrations type of work.
We don't have anything clearly defined at the moment, but we should have it better defined in the weeks following the on site.

brunoguerios commented on modules/sor/sorV2/lib/path.ts:
```diff
@@ -44,12 +56,29 @@ export class PathWithAmount extends PathLocal {
-                        this.mutateBalances,
-                    );
+                    let outputAmount;
+                    if (this.operations && this.operations[i] === PathOperation.AddLiquidity) {
```
Instead of adding this extra operations concept, I'd go with something more in line with SC implementation, where you check for each step:
- tokenOut === pool.address -> addLiquidity
- tokenIn === pool.address -> removeLiquidity
- else -> swap

brunoguerios commented on modules/sor/sorV2/lib/pools/composableStable/composableStablePool.ts:
```diff
@@ -345,4 +345,58 @@ export class ComposableStablePool implements BasePool {
+        amount: TokenAmount,
+        mutateBalances: boolean,
+    ): TokenAmount {
+        return this.swapGivenIn(tokenIn, bpt, amount, mutateBalances);
```
This is actually not possible for any pool on v3, since none of them contain its BPT as a token. You should be able to reuse the implementation you added to Weighted pools.

brunoguerios commented on modules/sor/sorV2/lib/pools/weighted/weightedPool.ts:
```diff
@@ -105,15 +106,15 @@ export class WeightedPool implements BasePool {
     public getNormalizedLiquidity(tokenIn: Token, tokenOut: Token): bigint {
-        const { tIn, tOut } = this.getRequiredTokenPair(tokenIn, tokenOut);
-
+        // const { tIn, tOut } = this.getRequiredTokenPair(tokenIn, tokenOut);
```
You should be able to remove this comment.

brunoguerios commented on modules/sor/sorV2/lib/pools/weighted/weightedPool.ts:
```diff
@@ -175,6 +188,83 @@ export class WeightedPool implements BasePool {
+    public addLiquiditySingleTokenExactIn(tokenIn: Token, bpt: Token, amount: TokenAmount): TokenAmount {
+        try {
+            // balances and amounts must be normalized to 1e18 fixed point - e.g. 1USDC => 1e18 not 1e6
+            const { tokenBalances, amountsIn, weights } = Array.from(this.tokenMap.values()).reduce(
```
Using `reduce` to "map" an array is usually a bad practice. You should be able to achieve the same result with `map`, which is much easier to follow/understand.

brunoguerios commented on modules/sor/sorV2/lib/pools/weighted/weightedPool.ts:
```diff
@@ -175,6 +188,83 @@ export class WeightedPool implements BasePool {
+            // balances and amounts must be normalized to 1e18 fixed point - e.g. 1USDC => 1e18 not 1e6
+            const { tokenBalances, amountsIn, weights } = Array.from(this.tokenMap.values()).reduce(
+                (acc: { tokenBalances: bigint[]; amountsIn: bigint[]; weights: bigint[] }, weightedPoolToken) => {
+                    if (weightedPoolToken.token.address.toLowerCase() === tokenIn.address.toLowerCase()) {
```
You should be able to use `isSameAddress`:
```suggestion
                    if (weightedPoolToken.token.isSameAddress(tokenIn.address) {
```

brunoguerios commented on modules/sor/sorV2/lib/pools/weighted/weightedPool.ts:
```diff
@@ -175,6 +188,83 @@ export class WeightedPool implements BasePool {
+
+    public removeLiquiditySingleTokenExactIn(tokenOut: Token, bpt: Token, amount: TokenAmount): TokenAmount {
+        const { amount: tokenBalance, weight: tokenWeight } = Array.from(this.tokenMap.values()).find(
+            (weightedPoolToken) => isSameAddress(weightedPoolToken.token.address, tokenOut.address),
```
You should be able to use isSameAddress from within Token:
```suggestion
            (weightedPoolToken) => weightedPoolToken.token.isSameAddress(tokenOut.address),
```

brunoguerios commented on modules/sor/sorV2/lib/pools/weighted/weightedPool.ts:
```diff
@@ -175,6 +188,83 @@ export class WeightedPool implements BasePool {
+    }
+
+    removeLiquiditySingleTokenExactOut(tokenOut: Token, bpt: Token, amount: TokenAmount): TokenAmount {
+        const { tokenBalances, amountsOut, weights } = Array.from(this.tokenMap.values()).reduce(
```
Same here about `reduce`

brunoguerios commented on modules/sor/sorV2/lib/pools/fx/fxPool.ts:
```diff
@@ -231,4 +231,38 @@ export class FxPool implements BasePool {
     }
+
+    addLiquiditySingleTokenExactIn(tokenIn: Token, bpt: Token, amount: TokenAmount): TokenAmount {
+        throw new Error('FX - addLiquiditiySingleTokenExactIn: Invalid Method');
```
As I mentioned, all pools should have all add/remove liquidity types available out of the box on v3.
I'm not really sure how you're going to handle that in the context of replicating pool math, since these methods were not available in v2 for these pools.
We'll likely have to replicate the `invariant` solution implemented by the SC team.

brunoguerios reviewed (changes requested):
I added a few comments, but I wasn't able to fully review the math side of things. 
I'll spend a bit more time on it in the following days, but I think you have enough feedback to work on the next few days.

lgahdl commented:
> Thanks for the PR. A few good catches and fixes in there :)
> 
> General remark:
> 
> 1. The idea was, that if one does not supply a vault version, it would query both v2 and v3 and return the better paths. Can you adapt to code for that or do we plan to add it later?
> 2. I would try to do the tokenPair data fetching in one go, whether they are BPT pairs or token Pairs.
>    ~3) Can you add poolToken->Phantom BPT pairs for CSP for v2? As @brunoguerios found out, there is a bug. CSP v2 pools support swap from pool token to phantom BPT.~ Bruno will add it to his PR
> 3. @brunoguerios a lot of pool math made its way into the API now. I wonder if we should start the project of the "pool math repo".
> 
> I havent reviewed all the math stuff, did you review that @brunoguerios ?
> 
> Did you run any queries and checked if it works? @brunoguerios @lgahdl 😅

answering about the queries, I ran with the queryFirst flag enabled, worked fine

lgahdl committed:
Update modules/sor/sorV2/lib/pools/weighted/weightedPool.ts

Co-authored-by: Bruno <bruno@balancerlabs.dev>

lgahdl committed:
Update modules/sor/sorV2/lib/pools/weighted/weightedPool.ts

Co-authored-by: Bruno <bruno@balancerlabs.dev>

Commit:
adding vaultVersion to tokenPairs;
generating bpt token pairs along with normal token pairs;
removing comment;
adding filter for pools with liquidity gte 1000;

Commit:
removing operation from pathGraphEdgeData and PathLocal and adapting to verify it comparing tokenIn/tokenOut with pool address;

Commit:
Comparing V2 with V3 and returning best result;

Commit:
Removing operations from PathWithAmount class;

Commit:
Removing operations from swapLocal and calculatePriceImpact;

Commit:
Merge commit '751bee44eff5e02b391e50818d718ff0b010b998' into v3-sor-with-swaps

Commit:
Adding Stable Pools with add/remove liquidity calculations for V3;
Moving Stable math to stable pool folder;
Adding Stable Pools to SOR;
Changing pool type compatible with add/remove liquidity paths from Composable Stable to Stable;

lgahdl commented:
As discussed with @brunoguerios I left the filter for STABLE and WEIGHTED pools, we will have a separated task to add the V3 Vault maths with invariant, and _computeInvariant and _computeBalance for all pool types

brunoguerios commented on modules/sor/sor.service.ts:
```diff
@@ -247,6 +267,37 @@ export class SorService {
+            return swapPath2;
+        } else if (swapPath2.returnAmount === '0') {
+            return swapPath1;
+        }
```
I know you can benefit from returning early, but since this check is quite simple, I'd say simpler is better:
```suggestion
```

brunoguerios commented on modules/sor/sorV2/lib/pools/gyro2/gyro2Pool.ts:
```diff
@@ -229,4 +229,38 @@ export class Gyro2Pool implements BasePool {
     }
+
+    addLiquiditySingleTokenExactIn(tokenIn: Token, bpt: Token, amount: TokenAmount): TokenAmount {
+        throw new Error('Gyro2 - addLiquiditiySingleTokenExactIn: Invalid Method');
```
Ok, so following up on what we discussed at the on-site, we'll stick with only Weighted/Stable pools on this first iteration, right? So please remove all other pool types.
We'll still go with previous pool math structure for this PR, but the idea is to setup the V3 math structure on a following PR and use it for all V3 operations, since it should be easier to add new pool types.
Later on, we should be able to evaluate if it's ok to move everything to the new math structure as ideally it should yield the same result.

brunoguerios commented on modules/sor/sorV2/lib/pools/gyro2/gyro2Pool.ts:
```diff
@@ -229,4 +229,38 @@ export class Gyro2Pool implements BasePool {
         return { tIn, tOut, sqrtAlpha, sqrtBeta };
     }
+
+    addLiquiditySingleTokenExactIn(tokenIn: Token, bpt: Token, amount: TokenAmount): TokenAmount {
```
Ok, so following up on what we discussed at the on-site, we'll stick with only Weighted/Stable pools on this first iteration, right?
We'll still go with previous pool math structure for this PR, so I think you'll need to keep these placeholder methods as an "intermediary state" for now.
The idea is to setup the V3 math structure on a following PR and use it for all V3 operations.
I'm not sure how the implementation will end up looking like, but I'd assume that at that point you'll be able to remove these placeholder methods to cleanup the code related to V2 math.
Later on, we should be able to evaluate if it's ok to move everything to the new math structure as ideally it should yield the same result.

brunoguerios commented:
@lgahdl - I think the best way to test the implementation is to compare SOR results with on-chain queries to see if they're yielding similar results. Can you please focus on that once you're back from you time off?

lgahdl commented:
> @lgahdl - I think the best way to test the implementation is to compare SOR results with on-chain queries to see if they're yielding similar results. Can you please focus on that once you're back from you time off?

@brunoguerios Of course! This will be my focus, I made some tests before the last changes, I'll see if after the last changes the results kept good, thanks man!

lgahdl committed:
Update modules/sor/sor.service.ts

Co-authored-by: Bruno <bruno@balancerlabs.dev>

brunoguerios commented on modules/sor/sorV2/lib/pools/stable/stablePool.ts:
```diff
@@ -0,0 +1,364 @@
+        }
+    }
+
+    getLimitAmountAddLiquidity(tokenIn: Token): bigint {
```
It feels weird to have a limit for add liquidity amounts 🤔 
Where did you get this from?

brunoguerios commented on modules/sor/sorV2/lib/pools/weighted/weightedPool.ts:
```diff
@@ -131,6 +131,18 @@ export class WeightedPool implements BasePool {
+    }
+
+    public getLimitAmountRemoveLiquidity(): bigint {
+        return (this.totalShares * this.MAX_IN_RATIO) / WAD;
```
Also not sure if this number makes sense. Did you confirm this info?

lgahdl replied:
I didn't confirm this number, but I understood that this limit is necessary to prevent the pool's invariant of suffering a big change with swaps/adds/removes that exceed 30% of the balance of the token to be transferred, Tom French made this implementation back in the old SOR, I'll confirm if this makes sense with the SC guys

brunoguerios replied:
I understand the limit on swaps as this is something implemented on each pool contract and also in the previous SOR. 
My question is more about this limit on removeLiquidity. I don't recall seeing that on the old SOR. 
I understand that removing liquidity to a single token should have some sort of limit, as it behaves similarly to swapping, but I'm not sure about the math here. That's why I'm asking where did you get this math from.

lgahdl replied:
I applied the logic of a swap with BPT for RemoveLiquidity, if a swap with BPT had the limit of 30% of the totalSupply, I thought the RemoveLiquidity with a single token out should have the same limit, but maybe I'm missing something here?

brunoguerios replied:
I don't think it's as straightforward as that. If you take the stable pool swap limit for example, it's approximately the whole token balance. In the context of removeLiquidity, you wouldn't be able to remove liquidity to a single token with nearly all BPT balance.
As this is actually a limit stablished by the pool contract, I'd double check with SC team on what would be the limits for single token add/remove operations.
When asking them, don't forget that this is in the context of v3 only.

lgahdl replied:
Yeah, you're right, there is no limit in the V3, not even for swaps actually, I'll change the flow, so V3 pools will not have to calculate the limit, thanks

brunoguerios replied:
Even though the pool contract doesn't enforce a limit, I believe the SOR needs to implement some sort of limit anyway. Otherwise you won't be able to filter out pools that are too small to route the swap. 
If that's the case, I'd go with something like this:
- swaps: set the limit to the whole token balance
- addLiquidity: no limit
- removeLiquidity exactOut: limit is the amount of tokenOut
- removeLiquidity exactIn: you'll have to calculate some sort of pulled limit, by the available amount of tokenOut

Let me know if that makes sense.

lgahdl replied:
Now that I'm making the tests for these limits, I saw that this remove liquidity limits are not possible, to remove all the amount of one of the tokens, the user would have to send the invariant of the pool to 0, it breaks the current math. Maybe keep the limit as 30% of the amount of tokenOut is a better option in the moment? I'll rollback to 30% and then we can change it in the future.

brunoguerios commented on modules/sor/sorV2/lib/pathGraph/pathGraph.ts:
```diff
@@ -457,38 +506,75 @@ export class PathGraph {
-            const poolLimitExactIn = path[i].pool.getLimitAmountSwap(
-                path[i].tokenIn,
-                path[i].tokenOut,
+    private getLimitAmountSwapForPath(path: PathGraphEdgeData[]): bigint {
```
I'm confused with this `getLimitAmountSwapForPath` method.
Even the previous implementation seems to be comparing amountsIn with amountsOut, which doesn't make sense to me 😕 
Do you mind explaining how it works, so I can understand the changes you made?

lgahdl replied:
No, of course not, this method finds the limit amount of the input/output token for the path, this limit is marked in a way that no token/bpt balance of any pool of the path will be changed more than 30%

franzns commented on modules/sources/contracts/fetch-tokenpair-data.ts:
```diff
@@ -186,9 +199,48 @@ function generateTokenPairs(filteredPools: PoolInput[]): TokenPair[] {
+                poolTvl: pool.dynamicData?.totalLiquidity || 0,
+                // remove pools that have <$1000 TVL or a token without a balance or USD balance
+                valid:
+                    // V2 Validation
```
remove comment

franzns commented on package.json:
```diff
@@ -23,7 +23,7 @@
         "@aws-sdk/client-secrets-manager": "^3.195.0",
         "@aws-sdk/client-sqs": "^3.137.0",
         "@balancer-labs/sdk": "github:beethovenxfi/balancer-sdk#beethovenx-master",
-        "@balancer/sdk": "^0.12.0",
```
did you check whether this breaks anything else? Not quite sure where the old sdk is used.

lgahdl replied:
I did check, the core of the SDK is used only in the SOR, there are some of it being used on the fetch-tokenpair-data and pool-on-chain-tokenpair-data, but it's only constants(WAD and ZERO_ADDRESS) and MathSol, that didn't suffer any changes in these latest updates

brunoguerios commented on modules/sor/sorV2/lib/pathGraph/pathGraph.ts:
```diff
@@ -457,38 +506,75 @@ export class PathGraph {
+                    const pulledLimit: bigint = path[i].pool.addLiquiditySingleTokenExactOut(
+                        path[i].tokenIn,
+                        path[i].tokenOut,
+                        TokenAmount.fromRawAmount(path[i].tokenIn, limit),
```
It should be tokenOut here, right?

lgahdl replied:
yes, thanks!

franzns commented on modules/sor/sorV2/sorPathService.ts:
```diff
@@ -30,11 +30,15 @@ import {
 import { PathWithAmount } from './lib/path';
 import { calculatePriceImpact, getInputAmount, getOutputAmount } from './lib/utils/helpers';
 import { SwapLocal } from './lib/swapLocal';
+import { update } from 'lodash';
```
remove

franzns commented on modules/sor/sorV2/lib/static.ts:
```diff
@@ -37,6 +39,8 @@ export async function sorGetSwapsWithPools(
             case 'META_STABLE':
                 basePools.push(MetaStablePool.fromPrismaPool(prismaPool));
                 break;
+            case 'STABLE':
+                basePools.push(StablePool.fromPrismaPool(prismaPool));
```
this is actually the legacy stable pools, do we need them here?

lgahdl replied:
I thought of using it for the new Stable pools, but maybe it'll be better to create a new pool type for Stable V3 pools now that the math will change, the name that the API will give for V3 Stable pool type will be 'STABLE'?
Also, in the old V2 SOR the Stable Pools were included, I believe that they should be included here as well, even if I create another StablePool class for V3 pools.

brunoguerios commented on modules/sor/sorV2/lib/pathGraph/pathGraph.ts:
```diff
@@ -457,38 +506,75 @@ export class PathGraph {
+                    const pulledLimit: bigint = path[i].pool.removeLiquiditySingleTokenExactOut(
+                        path[i].tokenOut,
+                        path[i].tokenIn,
+                        TokenAmount.fromRawAmount(path[i].tokenIn, limit),
```
Same here - if it's exactOut, I'd expect tokenOut to be provided.

franzns reviewed (commented):
Thanks. Just a few minor comments to be addressed! 
Can you also add the test script you used to find the swaps?

brunoguerios commented on modules/sor/sorV2/lib/pathGraph/pathGraph.ts:
```diff
@@ -457,38 +506,75 @@ export class PathGraph {
+                        path[i].tokenOut,
+                        TokenAmount.fromRawAmount(path[i].tokenOut, limit),
+                    ).amount;
+                    limit = pulledLimit > limitGivenIn ? limitGivenOut : pulledLimit;
```
That's what was on the previous code. It makes more sense to me. Did you change it intentionally?
```suggestion
                    limit = pulledLimit > limitGivenIn ? limitGivenIn : pulledLimit;
```

brunoguerios commented on modules/sor/sorV2/lib/pools/basePool.ts:
```diff
@@ -10,5 +10,31 @@ export interface BasePool {
+    ): TokenAmount;
+    addLiquiditySingleTokenExactOut(
+        tokenIn: Token,
+        bpt: Token,
```
Do you really need `bpt` as input here? Can you please review these interfaces to make sure they have only the inputs they really need?

brunoguerios reviewed (changes requested):
I looked a bit more in depth to the implementation and found some issues that I believe would be easily spotted with tests.
If things like these only showed up in my 3rd review, I don't have much confidence that there isn't more similar issues. 😕
I know it's too big of a task to add tests for the whole SOR, but we should start somewhere.
Maybe just for this limit calculation at first?
What do you think?

Commit:
Removing comments;
Reviewing interfaces;
Fixing wrong parameters;

Commit:
Fixing operation interface usage;

Commit:
Reformulating limits for removeLiquidity;
Removing addLiquidity limits;

Commit:
Adding Tests;
Fixing PriceRate multiplier on stable pools;
Renaming test utils file to network-utils and adding another file (The network utils was throwing an error related with networkConfig file)
Creating factory for prisma pools and tokens;

brunoguerios commented on modules/sor/sorV2/lib/pools/weighted/weightedPool.ts:
```diff
@@ -124,11 +128,48 @@ export class WeightedPool implements BasePool {
+            throw new Error('getLimitRemoveLiquidity: Token not found');
+        }
+        if (removeLiquidityKind === RemoveLiquidityKind.SingleTokenExactOut) {
+            return (tOut.amount * this.MAX_OUT_RATIO) / WAD;
```
I'm still not sure about some of these limits. On this case, I expect a user to be able to remove liquidity with more than just `(tOut.amount * this.MAX_OUT_RATIO)`.

How about we create a doc on Notion with all these limits specified by:
- vault version (2 or 3)
- pool type (weighted, stable, etc..)
- operation (swap, add liquidity single token, remove liquidity single token)
- exact in/out

Then we'll have a centralized source of truth that everyone can easily check and we also make sure that it was validated by the SC team.

brunoguerios commented on modules/sor/sorV2/lib/pathGraph/pathGraph.ts:
```diff
@@ -457,38 +506,74 @@ export class PathGraph {
+    private getLimitAmountSwapForPath(path: PathGraphEdgeData[]): bigint {
+        let limit;
+        if (this.getOperationFromPath(path[path.length - 1]) === PathOperation.AddLiquidity) {
+            limit = MAX_UINT256;
```
Since it's a single sided Add Liquidity operation, there might be a limit. See other comment for suggestion on how to validate this.
I'd use `getLimitAmountAddLiquidity` anyway and get the amount from there.

brunoguerios commented on modules/sor/sorV2/lib/pathGraph/pathGraph.ts:
```diff
@@ -457,38 +506,74 @@ export class PathGraph {
+                limitGivenIn = path[i].pool.getLimitAmountSwap(path[i].tokenIn, path[i].tokenOut, SwapKind.GivenIn);
+                limitGivenOut = path[i].pool.getLimitAmountSwap(path[i].tokenIn, path[i].tokenOut, SwapKind.GivenOut);
+                if (limitGivenOut <= limit) {
+                    limit = limitGivenOut;
```
This is how it was implemented in the original code.
```suggestion
                    limit = limitGivenIn;
```

brunoguerios commented on modules/sor/sorV2/lib/pathGraph/pathGraph.ts:
```diff
@@ -457,38 +506,74 @@ export class PathGraph {
+                    RemoveLiquidityKind.SingleTokenExactOut,
+                );
+                if (limitGivenOut <= limit) {
+                    limit = limitGivenOut;
```
Should be `given in` here as well.
```suggestion
                    limit = limitGivenIn;
```

brunoguerios commented on modules/sor/sorV2/lib/pathGraph/pathGraph.ts:
```diff
@@ -457,38 +506,74 @@ export class PathGraph {
+                    limit = pulledLimit > limitGivenIn ? limitGivenIn : pulledLimit;
+                }
             }
         }
```
You'll need an extra `if` clause for `PathOperation.AddLiquidity`

brunoguerios reviewed (changes requested):
Nice that we have tests for `limits` now 👏 
I think we need to find a way to test/validate the math for swaps with add/remove liquidity steps as well.
Any ideas on how to do it?

Commit:
removing add/remove liquidity from SOR;

Commit:
removing unused variable

Commit:
Adding anvil setup;
Adding tests comparing query results with SOR results;

brunoguerios commented:
Closing this as duplicate #398
